libkbfs/bare_root_metadata_v3.go

```diff
@@ -511,6 +511,75 @@ func (md *BareRootMetadataV3) TlfHandleExtensions() (
 	return extensions
 }
 
+// PromoteReader implements the BareRootMetadata interface for
+// BareRootMetadataV3.
+func (md *BareRootMetadataV3) PromoteReader(
+	uid keybase1.UID, extra ExtraMetadata) error {
+	if md.TlfID().IsPublic() {
+		return InvalidPublicTLFOperation{md.TlfID(), "PromoteReader"}
+	}
+
+	wkb, rkb, ok := getKeyBundlesV3(extra)
+	if !ok {
+		return errors.New("Key bundles missing")
+	}
+	dkim, ok := rkb.Keys[uid]
+	if !ok {
+		return fmt.Errorf("Could not find %s in rkb", uid)
+	}
+	// TODO: This is incorrect! Fix this.
+	wkb.Keys[uid] = dkim
+	delete(rkb.Keys, uid)
+	return nil
+}
+
+// RevokeDevices implements the BareRootMetadata interface for
+// BareRootMetadataV3.
+func (md *BareRootMetadataV3) RevokeDevices(
+	keys []kbfscrypto.CryptPublicKey, extra ExtraMetadata) error {
+	if md.TlfID().IsPublic() {
+		return InvalidPublicTLFOperation{md.TlfID(), "RevokeDevices"}
+	}
+
+	wkb, rkb, ok := getKeyBundlesV3(extra)
+	if !ok {
+		return errors.New("Key bundles missing")
+	}
+
+	for _, key := range keys {
+		for _, dkim := range wkb.Keys {
+			delete(dkim, key)
```

Comment:
Oh, I see the answer in a `KeyManager` TODO now.  Maybe worth a comment/TODO here as well, as it does something different than v2.  The interface comment for `RevokeDevices` might need to be tweaked as well to specify more exactly what the implementer should do with historic key gens.